apps/prairielearn/src/sprocs/check_assessment_access_rule.sql

```diff
@@ -45,103 +43,48 @@ BEGIN
 
     << schedule_access >>
     DECLARE
-        ps_course_id bigint;
-        reservation reservations;
+        exam_uuid uuid;
     BEGIN
-        -- is an exam_id hardcoded into the access rule? Check that first
-        IF assessment_access_rule.exam_uuid IS NOT NULL THEN
-
-            -- require exam mode
-            IF check_assessment_access_rule.mode IS DISTINCT FROM 'Exam' THEN
-                authorized := FALSE;
-                EXIT schedule_access;
-            END IF;
-
-            -- is there a checked-in reservation?
-            SELECT r.*
-            INTO reservation
-            FROM
-                reservations AS r
-                JOIN exams AS e USING(exam_id)
-            WHERE
-                e.uuid = assessment_access_rule.exam_uuid
-                AND r.user_id = check_assessment_access_rule.user_id
-                AND r.delete_date IS NULL
-                AND date BETWEEN r.access_start AND r.access_end
-            ORDER BY r.access_end DESC -- choose the longest-lasting if >1
-            LIMIT 1;
-
-            IF FOUND THEN
-                -- we have a valid reservation, don't keep going to "authorized := FALSE"
-                EXIT schedule_access;
-            END IF;
+        -- error case, exam_uuids only work for mode:Exam assessments
+        IF assessment_access_rule.exam_uuid IS NOT NULL AND mode != 'Exam' THEN
+            authorized := FALSE;
+        END IF;
 
-            -- is there a checked-in pt_reservation?
-            SELECT r.access_end
-            INTO exam_access_end
-            FROM
-                pt_reservations AS r
-                JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
-                JOIN pt_exams AS x ON (x.id = r.exam_id)
-            WHERE
-                (date BETWEEN r.access_start AND r.access_end)
-                AND e.user_id = check_assessment_access_rule.user_id
-                AND x.uuid = assessment_access_rule.exam_uuid;
+        -- only enforce for mode:Exam otherwise skip
+        IF mode != 'Exam' THEN
+            EXIT schedule_access;
+        END IF;
 
-            IF FOUND THEN
-                -- we have a valid reservation, don't keep going to "authorized := FALSE"
+        -- is there a checked-in pt_reservation?
+        SELECT x.uuid
+        INTO exam_uuid
+        FROM
+            pt_reservations AS r
+            JOIN pt_enrollments AS e ON (e.id = r.enrollment_id)
+            JOIN pt_exams AS x ON (x.id = r.exam_id)
+        WHERE
+            (date BETWEEN r.access_start AND r.access_end)
+            AND e.user_id = check_assessment_access_rule.user_id;
+
+        IF FOUND THEN
+            IF assessment_access_rule.exam_uuid = exam_uuid THEN
+                -- exam_uuid matches, so don't keep going to authorized := FALSE
                 EXIT schedule_access;
+            ELSE
+                -- checked-in so deny any exams that is not linked
+                authorized := FALSE;
             END IF;
```

Comment:
I just tested, and it is possible for a student to have more than one active reservation at a time, so I assume the exam_uuid will need to go to the query instead of to a separate test.